Parse signatures when collecting referenced nested classes

lrytz opened the pull request:
To emit the InnerClass table we visit ClassNodes after all code
generation is done. This is simpler than keeping track of things
during code generation and the optimizer.

Until now we didn't include references to nested classes that
appear only in generic signatures.

Fixes scala/bug#10180, scala/bug#10351, scala/bug#9810

lrytz committed:
Move ClassNode traverser for collecting nested classes

For easier testability

lrytz committed:
Fix generic signatures for classes nested in modules

lrytz committed:
Test for java inner class referenced in signature only

lrytz committed:
Don't emit generic signatures for types erasing to a primitive

Fields with type aliases erasing to a primitive should not
get a generic signature.

Fixes scala/bug#10351, scala/bug#9810

lrytz commented:
jardiff results look as expected
  - 76038d898e-1fada8e035 https://gist.github.com/lrytz/b3444099905d3f78cd01a95730ea2b79
    - fixes a signature
  - 1fada8e035-63da4bed24 https://gist.github.com/lrytz/8f3b22ed315a02e729438594e26d093d
    - no change in the generated code, there was only a test affected
  - 63da4bed24-c43fc72d7a https://gist.github.com/lrytz/0b8b8a158ebcdccbc3526ba72d5e9337
    - lots of new InnerClass entries, for example for class [ImplicitTags](https://github.com/scala/scala/blob/2.12.x/src/reflect/scala/reflect/api/ImplicitTags.scala)

lrytz commented:
Maybe we should wrap signature parsing in a try-catch-any for now. There were a number of crashes caused by bugs in our signature generation. If some user code triggers other bugs in signature generation that would cause a regression (backend crash). Started a community build here https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/1546/consoleFull

retronym commented:
👍 for the adding try/catch. Good idea to run the community build before we add that. Another way to test would be to run the signature parser across a wide selection of JARs from Maven Central (in particular, Scala originated ones)

retronym commented:
My version of the nested module signature fix is in https://github.com/scala/scala/compare/2.12.x...retronym:ticket/10180. Will review your code in more detail soon.

I also touched on signature generation recently in https://github.com/retronym/scala/commit/4a7eb3663081dc03a87185c102628c96cce0c36a, but that looks performance oriented only. I wanted to get rid of the `dotCleanup` part, which is a crutch for bugs in the code above it, but the whole things needed work to get to that point...

lrytz commented:
Aw, I didn't remember your PR.. At least some of the time I spent went into fixing some other related issues. What is your preference for handling errors: Silent, dev warning, warning?

retronym commented:
@lrytz Let's see how noisily `javac` warns about the same sort of problem and follow that.

lrytz committed:
Parse signatures when collecting referenced nested classes

To emit the InnerClass table we visit ClassNodes after all code
generation is done. This is simpler than keeping track of things
during code generation and the optimizer.

Until now we didn't include references to nested classes that
appear only in generic signatures.

Fixes scala/bug#10180

lrytz commented:
@retronym I created a blend of your PR (#5690) and mine, ready to review.

lrytz commented on src/compiler/scala/tools/nsc/transform/Erasure.scala:
```diff
@@ -282,21 +282,26 @@ abstract class Erasure extends InfoTransform
+                  // If the prefix is a module, drop the '$'. Classes (or modules) nested in modules
+                  // are separated by a single '$' in the filename: `object o { object i }` is o$i$.
+                  val withoutOwningModuleDollar =
+                    if (preRebound.typeSymbol.isModuleClass) withoutSemi.stripSuffix(nme.MODULE_SUFFIX_STRING)
```
In your version (https://github.com/scala/scala/pull/5690/commits/21b4378a04d404a2e2c7547b28ce2ee3d7f268fe#diff-34082ec02c067748f0d47c0bf037a717R292) there's an additional check `&& sym.isModuleOrModuleClass`, but that's not correct: we also need to do it for classes, see the last line in the test of this commit.

retronym commented:
Measuring the performance impact presently, data for the tip of this PR should [appear shortly](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?var-scalaSha=799bb89%7Ce5ea3ab&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40&from=1492395954042&to=1497579954042).

lrytz commented:
Benchmark results are up now, there seems to be some slowdown (3.5% for scalap, insignificant for vector / better-files).

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala:
```diff
@@ -543,3 +450,277 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
+          // skips over '[', ')', primitives
+          i += 1
+        }
+        internalNames foreach visitInternalName
```
Can we avoid building up the collection here and just directly call `visitInternalName` immediately after we find each class name?

lrytz replied:
good catch, that must be an artefact of iterative development

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala:
```diff
@@ -543,3 +450,277 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
+            while (desc.charAt(i) != ';') i += 1
+            internalNames += desc.substring(start, i)
+          }
+          // skips over '[', ')', primitives
```
`'^'` as well?

retronym replied:
I guess not, I'm thinking of signatures, not descriptors.

lrytz replied:
yep

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala:
```diff
@@ -543,3 +450,277 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
+
+      @tailrec private def referenceTypeSignature(): Unit = getCurrentAndSkip() match {
+        case 'L' =>
+          val names = new java.lang.StringBuilder()
```
Might pay to start this with an initial capacity bigger than the default of `16`, which is likely to be too small for a significant proportion of FQCNs.

lrytz replied:
64? for example `scala/collection/immutable/HashSet` is 34.

retronym replied:
I was tossing up between 32 and 64.

lrytz replied:
We could also benchmark the idea of keeping a single `StringBuilder` and using `setLength(0)` (https://stackoverflow.com/a/5193094/248998)

retronym replied:
Based on a [back of the envelope](https://docs.google.com/spreadsheets/d/1uzsap5f7UV-Z42mivUqk1Ly38cE5VvDGh4geMgsTzq0/edit?usp=sharing) analysis of SCALA_HOME/lib and a selection of JARs from my ivy cache, I'd say 64. would be preferable.

lrytz replied:
I did a quick benchmark (https://github.com/lrytz/compiler-benchmark/commits/nestedClassCollector) for the various ideas (revisions here https://github.com/lrytz/scala/commits/t10180-wip)

```
* e5ea3abfa3 - (origin/t10180) Parse signatures when collecting referenced nested classes (8 days ago) <Lukas Rytz>
# VM options: -DscalaVersion=2.12.3-bin-e5ea3ab -DscalaRef=e5ea3ab -Dsbt.launcher=/Users/luc/.sbt/launchers/0.13.15/sbt-launch.jar
NestedClassCollector.rtJar  thrpt   20  4.902 ± 0.104  ops/s

* a34aa97968 - unnecessary ListBuffer (2 hours ago) <Lukas Rytz>
# VM options: -DscalaVersion=2.12.3-bin-a34aa97 -DscalaRef=a34aa97 -Dsbt.launcher=/Users/luc/.sbt/launchers/0.13.15/sbt-launch.jar
NestedClassCollector.rtJar  thrpt   20  4.876 ± 0.116  ops/s

* 2132e98f13 - charBooleanFun (2 hours ago) <Lukas Rytz>
# VM options: -DscalaVersion=2.12.3-bin-2132e98 -DscalaRef=2132e98 -Dsbt.launcher=/Users/luc/.sbt/launchers/0.13.15/sbt-launch.jar
NestedClassCollector.rtJar  thrpt   20  5.146 ± 0.104  ops/s

* c8f7eb73b5 - initial size (2 hours ago) <Lukas Rytz>
# VM options: -DscalaVersion=2.12.3-bin-c8f7eb7 -DscalaRef=c8f7eb7 -Dsbt.launcher=/Users/luc/.sbt/launchers/0.13.15/sbt-launch.jar
NestedClassCollector.rtJar  thrpt   20  5.148 ± 0.101  ops/s

* 4b4f863c46 - (HEAD -> t10180-wip, origin/t10180-wip) single builder (2 hours ago) <Lukas Rytz>
# VM options: -DscalaVersion=2.12.3-bin-4b4f863 -DscalaRef=4b4f863 -Dsbt.launcher=/Users/luc/.sbt/launchers/0.13.15/sbt-launch.jar
NestedClassCollector.rtJar  thrpt   20  5.115 ± 0.129  ops/s
```

So the `CharBooleanFunction` helped, but then setting an initial size for the buffer here or using a single buffer didn't make a difference.

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala:
```diff
@@ -543,3 +450,277 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
+      private val Aborted: Throwable = new NoStackTrace { }
+      private def abort(): Nothing = throw Aborted
+
+      def safely(f: => Unit): Unit = try f catch {
```
I assume this is intended to be inlined, perhaps add the annotation to express the desire more strongly.

retronym commented on src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala:
```diff
@@ -543,3 +450,277 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
+
+      private def skipUntil(isDelimiter: Char => Boolean): Unit = {
+        while (!isDelimiter(current)) { index += 1 }
+      }
```
Use a SAM type `trait CharBooleanFunction { def apply(c: Char) => Boolean } ` to avoid boxing.

```
scala> ((c: Char) => true).getClass.getInterfaces
res2: Array[Class[_]] = Array(interface scala.Function1, interface scala.Serializable)
```

lrytz replied:
sure, good point!

retronym commented on test/junit/scala/tools/nsc/backend/jvm/NestedClassesCollectorTest.scala:
```diff
@@ -0,0 +1,115 @@
+      classNode.methods.iterator().asScala.map(_.signature).foreach(c.visitMethodSignature)
+      classNode.fields.iterator().asScala.map(_.signature).foreach(c.visitFieldSignature)
+    }
+  }
```
I ran this across the contents of my Ivy cache.

## Invalid signature

Needs investigation.

```
/Users/jz/.ivy2/cache/org.scala-sbt/org.scala-sbt-compiler-interface-0.13.13-bin_2.12.2-20170328-073517-729534b__52.0/jars/org.scala-sbt-compiler-interface-0.13.13-bin_2.12.2-20170328-073517-729534b__52.0-0.13.13_20161028T115300.jar

currentClass = /xsbt/Message.class
java.lang.Exception: Out of bounds, 63 >= 63 <T:Ljava/lang/Object;()>(Lscala/Function0<TT;>;)Lxsbti/F0<TT;>;
	at scala.tools.nsc.backend.jvm.Collector.$anonfun$raiseError$1(NestedClassesCollectorTest.scala:19)
	at scala.Option.getOrElse(Option.scala:121)
	at scala.tools.nsc.backend.jvm.Collector.raiseError(NestedClassesCollectorTest.scala:19)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.current(BackendUtils.scala:613)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.skipUntil(BackendUtils.scala:631)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.typeParameters(BackendUtils.scala:687)

```

```
% javap -cp /Users/jz/.ivy2/cache/org.scala-sbt/org.scala-sbt-compiler-interface-0.13.13-bin_2.12.2-20170328-073517-729534b__52.0/jars/org.scala-sbt-compiler-interface-0.13.13-bin_2.12.2-20170328-073517-729534b__52.0-0.13.13_20161028T115300.jar xsbt/Message
Compiled from "Message.scala"
public final class xsbt.Message {
Error: A serious internal error has occurred: java.lang.StringIndexOutOfBoundsException: String index out of range: -22
Please file a bug report, and include the following information:
java.lang.StringIndexOutOfBoundsException: String index out of range: -22
	at java.lang.String.substring(String.java:1967)
```

## Invalid signature due to value class wrapping

As fixed in this PR.

```
currentClass = /org/scalacheck/util/Buildable$$anon$5.class
/Users/jz/.ivy2/cache/org.scala-tools.testing/scalacheck_2.9.0-1/jars/scalacheck_2.9.0-1-1.9.jar	
scala.MatchError: > (of class java.lang.Character)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.referenceTypeSignature(BackendUtils.scala:658)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.typeArguments(BackendUtils.scala:653)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.referenceTypeSignature(BackendUtils.scala:664)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.classSignature(BackendUtils.scala:699)
cu
```

```
currentClass = /org/specs2/internal/scalaz/std/AnyValInstances$booleanInstance$disjunction$.class
/Users/jz/.ivy2/cache/org.specs2/scalaz-effect_2.10/jars/scalaz-effect_2.10-7.0.0.jar
scala.MatchError: Z (of class java.lang.Character)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.referenceTypeSignature(BackendUtils.scala:658)

/Users/jz/.ivy2/cache/org.specs2/specs2-scalaz-core_2.10/jars/specs2-scalaz-core_2.10-7.0.0.jar
scala.MatchError: Z (of class java.lang.Character)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.referenceTypeSignature(BackendUtils.scala:658)

```

## Corrupt class file? ASM bug?

```
/Users/jz/.ivy2/cache/com.ibm.icu/icu4j/jars/icu4j-2.6.1.jar
currentClass = /com/ibm/icu/impl/data/LocaleElements_zh__PINYIN.class
java.lang.ArrayIndexOutOfBoundsException: 48188
	at scala.tools.asm.ClassReader.readClass(ClassReader.java:2461)
	at scala.tools.asm.ClassReader.accept(ClassReader.java:542)
```

retronym replied:
> Needs investigation.

I'm almost certain that the errant classfile was built with a buggy compiler I created when trying to optimize the creation of generic signatures.

retronym replied:
FTR:

```scala
// for i in $(find /Users/jz/.ivy2/cache -name jars); do find $i -name '*.jar' | head -1; done > /tmp/jars
    val allJars = Files.readAllLines(Paths.get("/tmp/jars.txt")).asScala
    for (path <- allJars) {
      var currentClass: Path = null
      try {
        import java.nio.file._
        import scala.collection.JavaConverters._
        val zipfile = Paths.get(path)
        println(path)
        val fs = FileSystems.newFileSystem(zipfile, null)
        val root = fs.getRootDirectories.iterator().next()
        val contents = Files.walk(root).iterator().asScala.toList
        for (f <- contents if Files.isRegularFile(f) && f.getFileName.toString.endsWith(".class")) {
          currentClass = f
          val classNode = AsmUtils.classFromBytes(Files.readAllBytes(f))
          c.visitClassSignature(classNode.signature)
          classNode.methods.iterator().asScala.map(_.signature).foreach(c.visitMethodSignature)
          classNode.fields.iterator().asScala.map(_.signature).foreach(c.visitFieldSignature)
        }
      } catch {
        case t: Throwable =>
          println("currentClass = " + currentClass)
          t.printStackTrace()
      }
    }
```

lrytz replied:
no crash for my ivy repo when running this test. i'll add it as well.

retronym commented:
My micro-optimizations seem to speed up the signature parser a little.

```
  @Test
  //@Ignore("manually run test")
  def rtJar(): Unit = {
    import scala.collection.JavaConverters._
    // for i in $(find /Users/jz/.ivy2/cache -name jars); do find $i -name '*.jar' | head -1; done > /tmp/jars
    val allJars = Files.readAllLines(Paths.get("/tmp/jars.txt")).asScala.take(20)
    val contentss = for (path <- allJars) yield {
      import java.nio.file._

      val zipfile = Paths.get(path)
      val fs = FileSystems.newFileSystem(zipfile, null)
      val root = fs.getRootDirectories.iterator().next()
      val classFiles = Files.walk(root).iterator().asScala.toList.filter(f => Files.isRegularFile(f) && f.getFileName.toString.endsWith(".class"))
      classFiles.map(f => (f, AsmUtils.classFromBytes(Files.readAllBytes(f))))
    }
    var currentClass: Path = null
    for (i <- 1 to 2000) {
      for (contents <- contentss; (f, classNode) <- contents) {
        try {
          currentClass = f
          c.visitClassSignature(classNode.signature)
          classNode.methods.iterator().asScala.map(_.signature).foreach(c.visitMethodSignature)
          classNode.fields.iterator().asScala.map(_.signature).foreach(c.visitFieldSignature)
        }
        catch {
          case t: Throwable =>
            println("currentClass = " + currentClass)
            t.printStackTrace()
        }
      }
    }
}

```

Goes from 37s => 32s (albeit in an unscientific "benchmark")

retronym reviewed (approved):
This looks good to go, feel free to incorporate the micro-opts now or to wait until we look at this in the context of a larger profile.

lrytz commented:
> micro-opts

Are you referring to `CharBooleanFunction` and the other review comments?

retronym commented:
Yep. I think they are all beneficial, but I haven't measured properly.

lrytz committed:
Review feedback, some performance-related cleanups
